app/src/main/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenter.kt

```diff
@@ -0,0 +1,187 @@
+private const val TYPE_LOGIN_DEEP_LINK = 4
+
+class LoginOptionsPresenter @Inject constructor(
+    private val view: LoginOptionsView,
+    private val strategy: CancelStrategy,
+    private val factory: RocketChatClientFactory,
+    private val navigator: AuthenticationNavigator,
+    private val settingsInteractor: GetSettingsInteractor,
+    private val localRepository: LocalRepository,
+    private val saveCurrentServer: SaveCurrentServerInteractor,
+    private val saveAccountInteractor: SaveAccountInteractor,
+    private val analyticsManager: AnalyticsManager,
+    private val tokenRepository: TokenRepository,
+    serverInteractor: GetConnectingServerInteractor
+) {
+    // TODO - we should validate the current server when opening the app, and have a nonnull get()
+    private var currentServer = serverInteractor.get()!!
+    private lateinit var client: RocketChatClient
+    private lateinit var settings: PublicSettings
+    private lateinit var credentialToken: String
+    private lateinit var credentialSecret: String
+    private lateinit var deepLinkUserId: String
+    private lateinit var deepLinkToken: String
+    private lateinit var loginMethod: AuthenticationEvent
+
+    fun toCreateAccount() = navigator.toCreateAccount()
+
+    fun toLoginWithEmail() = navigator.toLogin(currentServer)
+
+    fun authenticateWithOauth(oauthToken: String, oauthSecret: String) {
+        credentialToken = oauthToken
+        credentialSecret = oauthSecret
+        loginMethod = AuthenticationEvent.AuthenticationWithOauth
+        doAuthentication(TYPE_LOGIN_OAUTH)
+    }
+
+    fun authenticateWithCas(casToken: String) {
+        credentialToken = casToken
+        loginMethod = AuthenticationEvent.AuthenticationWithCas
+        doAuthentication(TYPE_LOGIN_CAS)
+    }
+
+    fun authenticateWithSaml(samlToken: String) {
+        credentialToken = samlToken
+        loginMethod = AuthenticationEvent.AuthenticationWithSaml
+        doAuthentication(TYPE_LOGIN_SAML)
+    }
+
+    fun authenticateWithDeepLink(deepLinkInfo: LoginDeepLinkInfo) {
+        val serverUrl = deepLinkInfo.url
+        setupConnectionInfo(serverUrl)
+        if (deepLinkInfo.userId != null && deepLinkInfo.token != null) {
+            deepLinkUserId = deepLinkInfo.userId
+            deepLinkToken = deepLinkInfo.token
+            tokenRepository.save(serverUrl, Token(deepLinkUserId, deepLinkToken))
+            loginMethod = AuthenticationEvent.AuthenticationWithDeeplink
+            doAuthentication(TYPE_LOGIN_DEEP_LINK)
+        }
+    }
+
+    private fun doAuthentication(loginType: Int) {
+        launchUI(strategy) {
+            view.showLoading()
+            try {
+                val token = retryIO("login") {
+                    when (loginType) {
+                        TYPE_LOGIN_OAUTH -> client.loginWithOauth(credentialToken, credentialSecret)
+                        TYPE_LOGIN_CAS -> {
+                            delay(3, TimeUnit.SECONDS)
```

Comment:
Why we need this delay?

Reply:
We need to add a delay to make the CAS authentication work properly. If we remove that we'll see a message like "Unable to authenticate". This delay is needed due to the CAS authentication processing.